### PR TITLE
refactor: printToPDF should be headless

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

BREAKING CHANGE

Closes https://github.com/electron/electron/issues/24947.
Closes https://github.com/electron/electron/issues/30753.
Closes https://github.com/electron/electron/issues/27605.
Closes https://github.com/electron/electron/issues/29324.

This PR refactors our `printToPDF` implementation to match that of Chromium's headless implementation, seen [here](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/devtools/protocol/page_handler.h;l=54). Per https://bugs.chromium.org/p/chromium/issues/detail?id=1101596, we'd hit the limit with our previous approach, since with out of process iframes enabled we were unable to properly composite subframes. This removes that hacked-together implementation in favor of headless'.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Refactored `webContents.printToPDF` to align with the Chrome Devtools implementation.

### CODE REVIEW

**Comment by itsananderson** [comment]
API LGTM

**Review by itsananderson — APPROVED** [review]
Oops. I forgot how Github's review workflow works and didn't submit my comment.

**Comment by pushkin-** [comment]
@codebytere  By the way, I'm seeing a similar issue when using the `print` method instead of `printToPDF` on a window that's loading a PDF. Will that also be fixed by this PR, or should I create a new issue?

**Comment by codebytere** [comment]
@pushkin- that's a totally different problem & there are already open issues for it :)

**Comment by pushkin-** [comment]
> @pushkin- that's a totally different problem & there are already open issues for it :)

Cool, I found this [one](https://github.com/electron/electron/issues/30947). If that's the one you're talking about, can the "blocked" label be removed since there is a repro gist now? And can the version tag be updated? Thank you @codebytere

**Comment by codebytere** [comment]
To-Do
- Annotate code lifted from Chromium w/ location
- Update breaking changes doc

File: docs/api/web-contents.md
Comment (ckerr) [inline docs/api/web-contents.md:1477]: It looks like it wouldn't be that hard to honor the removed old options for some interim period, e.g. we could silently convert a `Record<string, number> pageRanges` to a string before passing it along to the new code.

_If_ that's the case and backwards support is moderately easy, the advantage is that this PR would no longer be a breaking change and it would give app devs more lead time to update their code.

The breaking changes doc could be updated to just say something like "these options will be removed in version X; please update to the new forms before then"

**Review by ckerr — COMMENTED** [review]
Full review forthcoming; just sending this one suggestion out for rn

**Review by jkleinsc — COMMENTED** [review]
API LGTM

**Review by ckerr — COMMENTED** [review]
API LGTM although if possible, backwards compatiability discussed [here](https://github.com/electron/electron/pull/33654#discussion_r872590695) would be nice-to-have

**Comment by codebytere** [comment]
Blocked on https://github.com/electron/electron/pull/34271

File: docs/api/web-contents.md
Comment (codebytere) [reply docs/api/web-contents.md]: Unfortunately it won't be possible to maintain backwards compat for all of the removed/altered ootions - meaning even if we maintained it for some (e.g. `pageRanges`) it'd be a `semver-major` breaking change regardless. Given that, i'm not sure it'd be particularly helpful to have piecemeal backwards compat 🤔  if you think otherwise i'm open to it though!

**Review by VerteDinde — APPROVED** [review]
I spent some time reviewing this yesterday on both Mac and Windows, using these two gists for testing ([one](https://gist.github.com/codebytere/bf98a7d34f4764ab1fcf2991fc4b35a8), [two](https://gist.github.com/codebytere/0f0e776b138452d6815ef4efc16aa0d1)). I didn't hit any errors, and the code looks good to me. Happy to do some smoke testing on Linux, but this looks good to me 👍

**Pull request merged**
